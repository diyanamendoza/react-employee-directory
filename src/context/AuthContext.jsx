import { createContext, useContext, useMemo, useState } from 'react'
import { getUser } from '../services/users'

const AuthContext = createContext()

const AuthProvider = ({ children }) => {
    const currentUser = getUser()
    const [user, setUser] = useState(currentUser ? { id: currentUser.id, email: currentUser.email } : '')
    const value = useMemo(() => ({ user, setUser }), [user])
    return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
}

const useAuth = () => {
    const context = useContext(AuthContext)
    if (context === undefined) { throw new Error('useAuth not used within Auth Provider')}
    return context
}

export { AuthContext, AuthProvider, useAuth }