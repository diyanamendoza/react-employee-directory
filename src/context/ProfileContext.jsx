import { createContext, useContext, useMemo, useState, useEffect } from 'react'
import { getProfile } from '../services/profiles'

export const ProfileContext = createContext()

export const ProfileProvider = ({ children }) => {
    const [profileStatus, setProfileStatus] = useState(false)
    const [profileData, setProfileData] = useState({}) 

    useEffect(() => {
        const hasProfile = async () => {
            try {
                const profileResult = await getProfile()
                if(profileResult.email) 
                    {
                    setProfileStatus(true)
                    setProfileData(profileResult)
                    }
            } catch (error) { console.log(error) }
        }
        hasProfile()
    }, [])

    const contextValue = useMemo(() => {
        return { profileStatus, profileData, setProfileData }
    }, [profileData])

    return (
        <ProfileContext.Provider value={contextValue}>{children}</ProfileContext.Provider>
    )
}

export const useProfile = () => {
    const context = useContext(ProfileContext)
    if (context === undefined) { throw new Error('useProfile hook not called inside provider')}
    return context
}