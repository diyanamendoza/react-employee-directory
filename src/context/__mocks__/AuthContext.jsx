//MOCK AUTH CONTEXT

import { createContext, useContext, useMemo, useState } from 'react'

const AuthContext = createContext()

const AuthProvider = ({ mockUser, children }) => {
    const [user, setUser] = useState(mockUser ? { id: mockUser.id, email: mockUser.email } : '')
    const value = useMemo(() => ({ user, setUser }), [user])
    return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
}

const useAuth = () => {
    const context = useContext(AuthContext)
    if (context === undefined) { throw new Error('useAuth not used within Auth Provider')}
    return context
}

export { AuthContext, AuthProvider, useAuth }