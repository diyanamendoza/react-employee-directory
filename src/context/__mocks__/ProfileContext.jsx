//MOCK PROFILE CONTEXT

import { createContext, useContext, useMemo, useState, useEffect } from 'react'

export const ProfileContext = createContext()

export const ProfileProvider = ({ children }) => {
    const [profileStatus, setProfileStatus] = useState(false)
    const [profileData, setProfileData] = useState({}) 

    useEffect(() => {
        const hasProfile = async () => {
            try {
                const profileResult = () => { 
                    return null
                }
                if(profileResult.email) 
                    {
                    setProfileStatus(true)
                    setProfileData(profileResult)
                    }
            } catch (error) { console.log(error) }
        }
        hasProfile()
    }, [])

    const contextValue = useMemo(() => {
        return { profileStatus, profileData, setProfileData }
    }, [profileData])

    return (
        <ProfileContext.Provider value={contextValue}>{children}</ProfileContext.Provider>
    )
}

export const useProfile = () => {
    const context = useContext(ProfileContext)
    if (context === undefined) { throw new Error('useProfile hook not called inside provider')}
    return context
}