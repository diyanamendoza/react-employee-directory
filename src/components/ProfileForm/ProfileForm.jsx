import React, { useState } from 'react'
import { useAuth } from '../../context/AuthContext'
import styles from '../../App.css'
import { useProfile } from '../../context/ProfileContext'
import { useHistory } from 'react-router-dom'



export default function ProfileForm({ handleProfile }) {
    const { user } = useAuth()
    const { profileData } = useProfile()
    const history = useHistory()
    

    const [fullname, setFullName] = useState(profileData.name || '')
    const [bio, setBio] = useState(profileData.bio || '')
    const [birthday, setBirthday] = useState(profileData.birthday || '')
    const [formError, setFormError] = useState('')

    const handleFormChange = (event) => {
      const { name, value } = event.target; 
      if(name === 'fullname') setFullName(value)
      if(name === 'bio') setBio(value)
      if(name === 'birthday') setBirthday(value)

    }

    const profileDataObj = {
      name: fullname,
      email: user.email,
      bio,
      birthday,
    }

    //incorporate handleProfile 
    const handleSubmit = async (event) => {
      event.preventDefault()
      try {
        await handleProfile(profileDataObj)
        history.replace('/')
      } catch (error) { 
        console.log(error)
        setFormError(error)
      }
    } 

    return (
        <div>
        <form className={styles.form} onSubmit={handleSubmit}>
        <fieldset className={styles.fieldset}>
          <legend>Your Profile</legend>
          <section className={styles.formSection}>
            <label htmlFor="fullname">Full Name</label>
            <input
            required
              id="fullname"
              type="name"
              name="fullname"
              value={fullname}
              className={styles.input}
              onChange={handleFormChange}
            />
          </section>
          <section className={styles.formSection}>
            <label htmlFor="email">Email</label>
            <input
            disabled
              id="email"
              type="email"
              name="email"
              value={user.email}
              className={styles.input}
              onChange={handleFormChange}
            />
          </section>
          <section className={styles.formSection}>
            <label htmlFor="birthday">Birthday</label>
            <input
            required
              id="birthday"
              type="date"
              name="birthday"
              value={birthday}
              className={styles.input}
              onChange={handleFormChange}
            />
          </section>
          <section className={styles.formSection}>
            <label htmlFor="bio">Bio</label>
            <textarea
            required
              id="bio"
              type="text"
              name="bio"
              value={bio}
              className={styles.input}
              onChange={handleFormChange}
            />
          </section>
          <button type='submit'>Submit</button>
          {formError && <p>{formError}</p>}
        </fieldset>
      </form>
        </div>
    )
}
