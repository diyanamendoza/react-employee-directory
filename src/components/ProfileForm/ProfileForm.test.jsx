import { render } from '@testing-library/react'
import ProfileForm from './ProfileForm'
import { MemoryRouter } from 'react-router-dom'
import { AuthProvider } from '../../context/AuthContext'
import { ProfileProvider } from '../../context/ProfileContext'

jest.mock('../../context/AuthContext')
jest.mock('../../services/users.js')
jest.mock('../../context/ProfileContext')

it('should display the profile form', () => {
    const { container } = render(
        <AuthProvider mockUser={{ id: 123, email: 'test@email.com'}}>
            <ProfileProvider>
             <MemoryRouter>
                <ProfileForm />
            </MemoryRouter>
            </ProfileProvider>
        </AuthProvider>
    )


    expect(container).toMatchSnapshot()
})