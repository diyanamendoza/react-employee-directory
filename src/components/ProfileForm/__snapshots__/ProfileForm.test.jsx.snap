// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the profile form 1`] = `
<div>
  <div>
    <form
      class="form"
    >
      <fieldset
        class="fieldset"
      >
        <legend>
          Your Profile
        </legend>
        <section
          class="formSection"
        >
          <label
            for="fullname"
          >
            Full Name
          </label>
          <input
            class="input"
            id="fullname"
            name="fullname"
            required=""
            type="name"
            value=""
          />
        </section>
        <section
          class="formSection"
        >
          <label
            for="email"
          >
            Email
          </label>
          <input
            class="input"
            disabled=""
            id="email"
            name="email"
            type="email"
            value="test@email.com"
          />
        </section>
        <section
          class="formSection"
        >
          <label
            for="birthday"
          >
            Birthday
          </label>
          <input
            class="input"
            id="birthday"
            name="birthday"
            required=""
            type="date"
            value=""
          />
        </section>
        <section
          class="formSection"
        >
          <label
            for="bio"
          >
            Bio
          </label>
          <textarea
            class="input"
            id="bio"
            name="bio"
            required=""
            type="text"
          />
        </section>
        <button
          type="submit"
        >
          Submit
        </button>
        
      </fieldset>
    </form>
  </div>
</div>
`;
